cpp/src/arrow/adapters/orc/adapter-test.cc

```diff
@@ -0,0 +1,161 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/adapters/orc/adapter.h"
+#include "arrow/array.h"
+#include "arrow/io/api.h"
+#include "orc/OrcFile.hh"
+
+#include <gtest/gtest.h>
+
+namespace liborc = orc;
+
+namespace arrow {
+
+constexpr int DEFAULT_MEM_STREAM_SIZE = 100 * 1024 * 1024;
+
+class MemoryOutputStream : public liborc::OutputStream {
+ public:
+  explicit MemoryOutputStream(ssize_t capacity) : name_("MemoryOutputStream") {
+    data_ = new char[capacity];
```

Comment:
Please either use Arrow's MemoryPool or `std::vector<char>` for this, then we would have automatic deletes in the destruction.